File: frontend/app/src/models/Review.js
/**
 * Review Model - Đại diện cho 1 đánh giá sân
 */
class Review {
  constructor(id, userId, venueId, rating, comment, authorName, createdAt, ownerReply) {
    this.id = id;
    this.userId = userId;
    this.venueId = venueId;
    this.rating = rating;        // 1 - 5
    this.comment = comment;
    this.authorName = authorName;
    this.createdAt = createdAt;
    this.ownerReply = ownerReply;
  }

  static fromAPI(data) {
    const authorName = data.userFullname ?? data.authorName ?? data.userName;
    const createdAt = data.createdAt ?? data.updatedAt;

    return new Review(
      data.id,
      data.userId,
      data.venueId,
      data.rating,
      data.comment,
      authorName,
      createdAt,
      data.ownerReply ?? null
    );
  }
}

export default Review;

File: backend/src/main/java/com/codewithvy/quanlydatsan/model/Review.java
package com.codewithvy.quanlydatsan.model;

import com.codewithvy.quanlydatsan.common.BaseEntity;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

/**
 * Entity lưu đánh giá của người dùng về một Venues sau khi hoàn thành booking.
 */
@Entity
@Table(name = "review")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class Review extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    // Người dùng đã đánh giá (USER)
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    // Venues được đánh giá
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "venue_id", nullable = false)
    private Venues venues;

    // Booking liên quan - đảm bảo chỉ review khi đã hoàn thành booking
    @OneToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "booking_id", nullable = false, unique = true)
    private Booking booking;

    // Đánh giá từ 1-5 sao
    @Column(nullable = false)
    private Integer rating;

    // Nhận xét văn bản
    @Column(columnDefinition = "TEXT")
    private String comment;

    // Phản hồi của chủ sân cho review
    @Column(name = "owner_reply", columnDefinition = "TEXT")
    private String ownerReply;

    // Ràng buộc: rating phải từ 1-5
    @PrePersist
    @PreUpdate
    private void validate() {
        if (rating < 1 || rating > 5) {
            throw new IllegalArgumentException("Rating must be between 1 and 5");
        }
    }
}

